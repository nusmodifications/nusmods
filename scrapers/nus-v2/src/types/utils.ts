/* eslint-disable import/prefer-default-export */

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>; // Remove types from T that are assignable to U
