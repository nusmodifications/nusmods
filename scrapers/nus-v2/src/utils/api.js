// @flow

/**
 * Small utility functions that don't need to be part of the main API class
 */

import { Logger } from 'bunyan';
import type { Semester } from '../types/modules';
import type { File } from '../components/fs';
import rootLogger from '../components/logger';

/**
 * Construct the 4 number term code from the academic year and semester
 */
export function getTermCode(semester: number | string, academicYear: string) {
  const year = /\d\d(\d\d)/.exec(academicYear);
  if (!year) throw new RangeError('academicYear should be in the format of YYYY/YYYY or YYYY-YY');
  return `${year[1]}${semester}0`;
}

/**
 * Extract the academic year and semester from a term code
 */
export function fromTermCode(term: string): [string, Semester] {
  const year = parseInt(term.slice(0, 2), 10);
  const semester = parseInt(term.charAt(2), 10);

  return [`20${year}/20${year + 1}`, semester];
}

export async function cacheDownload<T>(
  name: string,
  download: () => Promise<T>,
  cache: File<T>,
  logger: Logger = rootLogger,
): Promise<T> {
  try {
    // Try to download the data, and if successful cache it
    const data = await download();

    try {
      await cache.write(data);
    } catch (e) {
      logger.warn(`Failed to cache data to ${cache.path}`);
    }

    return data;
  } catch (e) {
    // If the file is not available we try to load it from cache instead
    logger.warn(e, `Cannot load ${name} from API, attempting to read from cache`);
    return cache.read();
  }
}
