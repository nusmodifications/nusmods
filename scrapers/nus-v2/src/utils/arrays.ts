/* eslint-disable import/prefer-default-export */

/**
 * Get the difference between each consecutive array elements. If there are
 * less than 2 elements in the array, an empty array is returned
 */
export function deltas(array: number[]): number[] {
  let previous = array[0];
  if (typeof previous === 'undefined') return [];

  const results = [];
  for (const next of array.slice(1)) {
    results.push(next - previous);
    previous = next;
  }

  return results;
}

export function allEqual<T>(array: T[]): boolean {
  const first = array[0];
  if (typeof first === 'undefined') return true;
  return array.every((element) => element === first);
}
