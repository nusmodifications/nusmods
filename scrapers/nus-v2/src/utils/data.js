// @flow

/**
 * Code for data mapping, cleanup and validation
 */

import { uniq, trim, groupBy, values } from 'lodash';
import type { VenueLesson } from '../types/venues';
import type { LessonWeek, ModuleCode } from '../types/modules';
import type { ModuleAliases } from '../types/mapper';

export const ZWSP = '\u200b';

/**
 * NUS specific title case function that accounts for school names, etc.
 */
const minorWords = ['and', 'or', 'the', 'a', 'an', 'to', 'of', 'in'];
const abbreviations = ['NUS', 'MIT', 'IP', 'NA'];
export function titleize(string: string) {
  let capitalized = string
    .toLowerCase()
    // http://stackoverflow.com/a/7592235
    .replace(/(?:^|\s\(?|-|\/)\S/g, (char) => char.toUpperCase());

  // Minor words are lowercase unless they are the first word of the title
  minorWords.forEach((properCasing) => {
    const regex = new RegExp(`\\b${properCasing}\\b`, 'gi');
    capitalized = capitalized.replace(regex, (match, index) =>
      index === 0 ? match : properCasing,
    );
  });

  // Abbreviations are uppercase regardless of where they are
  abbreviations.forEach((abbreviation) => {
    const regex = new RegExp(`\\b${abbreviation}\\b`, 'gi');
    capitalized = capitalized.replace(regex, abbreviation);
  });

  return capitalized;
}

/**
 * Remove keys with empty values, null or strings like 'nil', 'none'
 * Mutates the input object
 */
const trimChars = ' \t\n.,!?/-_';
const nullStrings = new Set(['', 'nil', 'na', 'n/a', 'null', 'none']);
export function removeEmptyValues<T: Object>(object: T, keys: $Keys<T>[]) {
  /* eslint-disable no-param-reassign */
  keys.forEach((key) => {
    const value = object[key];

    if (!value) delete object[key];
    if (typeof value === 'string' && nullStrings.has(trim(value, trimChars).toLowerCase())) {
      delete object[key];
    }
  });
  /* eslint-enable */

  return object;
}

/**
 * Trim given values if they are strings
 * Mutates the input object
 */
export function trimValues<T: Object>(object: T, keys: $Keys<T>[]) {
  /* eslint-disable no-param-reassign */
  keys.forEach((key) => {
    const value = object[key];

    if (typeof value === 'string') {
      object[key] = value.trim();
    }
  });
  /* eslint-enable */

  return object;
}

/**
 * Find modules which have lessons that fall on the exact same times
 */
export function getDuplicateModules(classes: VenueLesson[]): ModuleCode[] {
  const lessonsByTime: VenueLesson[][] = values(
    groupBy(classes, (lesson) => [lesson.StartTime, lesson.EndTime, lesson.Weeks, lesson.DayText]),
  );

  for (const lessons of lessonsByTime) {
    if (lessons.length > 1) {
      // Occasionally two classes share the same venue, so we don't count those
      const moduleCodes = uniq(lessons.map((lesson) => lesson.ModuleCode));
      if (uniq(moduleCodes).length > 1) return moduleCodes;
    }
  }

  return [];
}

/**
 * Remove duplicated lessons given by the modules array and merge their module
 * codes
 */
export function mergeModules(classes: VenueLesson[], modules: ModuleCode[]): VenueLesson[] {
  const mergedModuleCode = modules.join(`/${ZWSP}`);
  const removeModuleCodes = new Set(modules.slice(1));

  return classes
    .filter((lesson) => !removeModuleCodes.has(lesson.ModuleCode))
    .map((lesson) =>
      lesson.ModuleCode === modules[0]
        ? {
            ...lesson,
            ModuleCode: mergedModuleCode,
          }
        : lesson,
    );
}

export function mergeDualCodedModules(
  classes: VenueLesson[],
): { lessons: VenueLesson[], aliases: ModuleAliases } {
  let mergedModules = classes;
  let duplicateModules = getDuplicateModules(mergedModules);
  const aliases = {};

  while (duplicateModules.length) {
    // Mark the current set of modules as aliases
    for (const moduleCode of duplicateModules) {
      if (!aliases[moduleCode]) aliases[moduleCode] = new Set();
      duplicateModules
        .filter((moduleToAdd) => moduleToAdd !== moduleCode)
        .forEach((moduleToAdd) => aliases[moduleCode].add(moduleToAdd));
    }

    // Merge the lessons of the dual-coded module together
    mergedModules = mergeModules(mergedModules, duplicateModules);
    duplicateModules = getDuplicateModules(mergedModules);
  }

  return { lessons: mergedModules, aliases };
}

const weekOrder = {
  Orientation: 0,
  '1': 1,
  '2': 2,
  '3': 3,
  '4': 4,
  '5': 5,
  '6': 6,
  Recess: 6.5,
  '7': 7,
  '8': 8,
  '9': 9,
  '10': 10,
  '11': 11,
  '12': 12,
  '13': 13,
  Reading: 14,
};

export function compareWeeks(a: LessonWeek, b: LessonWeek) {
  return weekOrder[a] - weekOrder[b];
}

export const dayTextMap = {
  '1': 'Monday',
  '2': 'Tuesday',
  '3': 'Wednesday',
  '4': 'Thursday',
  '5': 'Friday',
  '6': 'Saturday',
  // Assume lessons on Sunday (7) are invalid
};

export const unrecognizedLessonTypes = {
  // Not recognized by frontend
  '4': 'Tutorial Type 4',
  '5': 'Tutorial Type 5',
  '6': 'Tutorial Type 6',
  '7': 'Tutorial Type 7',
  '8': 'Tutorial Type 8',
  '9': 'Tutorial Type 9',
  A: 'Supervision of Academic Exercise',
  O: 'Others',
  V: 'Lecture On Demand',
  I: 'Independent Study Module',
  C: 'Bedside Tutorial',
  M: 'Ensemble Teaching',
  J: 'Mini-Project',
};

export const activityLessonType = {
  // Recognized by frontend
  B: 'Laboratory',
  L: 'Lecture',
  D: 'Design Lecture',
  R: 'Recitation',
  P: 'Packaged Lecture',
  X: 'Packaged Tutorial',
  W: 'Workshop',
  E: 'Seminar-Style Module Class',
  S: 'Sectional Teaching',
  T: 'Tutorial',
  '2': 'Tutorial Type 2',
  '3': 'Tutorial Type 3',

  ...unrecognizedLessonTypes,
};
