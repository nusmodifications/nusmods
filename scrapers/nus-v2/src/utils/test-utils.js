// @flow

/* eslint-env jest */

import { omit, sortBy, zip } from "lodash";
import type { Cache } from "../services/output";
import type { Module, RawLesson, SemesterData } from "../types/modules";

/* eslint-disable import/prefer-default-export */

export function mockCache<T>(fileContent: T): Cache<T> {
  return {
    path: './fake',
    write: jest.fn().mockResolvedValue(),
    read: jest.fn().mockResolvedValue(fileContent),
  };
}

/**
 * Check the two objects have the same properties
 */
export function expectSameKeys<T: Object>(actual: T, expected: T) {
  expect(Object.keys(actual).sort()).toEqual(Object.keys(expected).sort());
}

export function serializeLesson(lesson: RawLesson) {
  return [
    lesson.LessonType,
    lesson.ClassNo,
    lesson.StartTime,
    lesson.WeekText,
    lesson.DayText,
    lesson.Venue,
  ].join('|');
}

/**
 * Check that two timetables are the same
 */
export function expectLessonsEqual(actual: RawLesson[], expected: RawLesson[]) {
  // Sort both expected and actual lessons because Jest expects
  // array to be in the same order
  expect(sortBy(actual, serializeLesson)).toEqual(sortBy(expected, serializeLesson));
}

function expectHistoryEqual(actual: SemesterData[], expected: SemesterData[]) {
  zip(
    sortBy(actual, (semester) => semester.Semester),
    sortBy(expected, (semester) => semester.Semester)
  ).forEach(([actualSemester, expectedSemester]) => {
    expect(omit(actualSemester, "Timetable")).toEqual(omit(expectedSemester, "Timetable"));

    expectLessonsEqual(actualSemester.Timetable, expectedSemester.Timetable);
  });
}

export function expectModulesEqual(actual: Module, expected: Module) {
  // LockedModules is excluded because the modules that require CS2100 are not part of this test
  const omittedKeys = ["History", "LockedModules"];
  expect(omit(actual, omittedKeys)).toEqual(omit(expected, omittedKeys));

  // Sort semesters and check history
  expectHistoryEqual(actual.History, expected.History);
}
