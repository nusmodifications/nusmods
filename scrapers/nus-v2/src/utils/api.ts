/**
 * Small utility functions that don't need to be part of the main API class
 */

import { Semester } from '../types/modules';
import { Cache } from '../services/io';
import rootLogger, { Logger } from '../services/logger';

/**
 * Construct the 4 number term code from the academic year and semester
 */
export function getTermCode(semester: number | string, academicYear: string) {
  const year = /\d\d(\d\d)/.exec(academicYear);
  if (!year) throw new RangeError('academicYear should be in the format of YYYY/YYYY or YYYY-YY');
  return `${year[1]}${semester}0`;
}

/**
 * Extract the academic year and semester from a term code
 */
export function fromTermCode(term: string): [string, Semester] {
  const year = parseInt(term.slice(0, 2), 10);
  const semester = parseInt(term.charAt(2), 10);

  return [`20${year}/20${year + 1}`, semester];
}

/**
 * Cache and return download if it succeeds, otherwise return cached data if
 * it has not expired yet
 */
export async function cacheDownload<T>(
  name: string,
  download: () => Promise<T>,
  cache: Cache<T>,
  logger: Logger = rootLogger,
): Promise<T> {
  try {
    // Try to download the data, and if successful cache it
    const data = await download();

    try {
      await cache.write(data);
    } catch (err) {
      logger.warn({ err, path: cache.path }, 'Failed to cache data');
    }

    return data;
  } catch (err) {
    // If the file is not available we try to load it from cache instead
    logger.warn(err, `Cannot load ${name} from API, attempting to read from cache`);
    return cache.read();
  }
}

/**
 * Retries the given promise
 */
export async function retry<T>(
  promiseFactory: () => Promise<T>,
  maxRetries: number,
  retryIf: (error: Error) => boolean = () => true,
): Promise<T> {
  try {
    return await promiseFactory();
  } catch (e) {
    // If we run out of tries, or if the given condition is not fulfilled, we
    // don't retry
    if (maxRetries <= 1 || !retryIf(e)) throw e;
    return retry(promiseFactory, maxRetries - 1, retryIf);
  }
}
