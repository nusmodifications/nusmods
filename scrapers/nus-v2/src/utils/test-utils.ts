import { AxiosRequestConfig, AxiosResponse } from 'axios';
import { omit, sortBy, zip } from 'lodash';
import httpStatus from 'http-status';

import { Cache } from '../services/io';
import { Weeks, Module, RawLesson, SemesterData } from '../types/modules';
import { Logger } from '../services/logger';

/* eslint-disable import/prefer-default-export */

export const EVERY_WEEK: Weeks = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13];

export function mockCache<T>(fileContent: T): jest.Mocked<Cache<T>> {
  // Annotating this as Cache<T> lets us make sure this function implements the interface
  const cache: Cache<T> = {
    path: './fake',
    write: jest.fn().mockResolvedValue(undefined),
    read: jest.fn().mockResolvedValue(fileContent),
  };

  // Returning this as jest.Mocked<Cache<T>> allows us to access .mock* functions
  return cache as any;
}

export function mockLogger(): Logger {
  return {
    // Mock all logged functions
    fatal: jest.fn(),
    error: jest.fn(),
    warn: jest.fn(),
    info: jest.fn(),
    debug: jest.fn(),
    trace: jest.fn(),

    // Calling child simply creates another mock logger
    child: mockLogger,
  };
}

/**
 * Check the two objects have the same properties
 */
export function expectSameKeys<T>(actual: T, expected: T) {
  expect(Object.keys(actual).sort()).toEqual(Object.keys(expected).sort());
}

export function serializeLesson(lesson: RawLesson) {
  return [
    lesson.LessonType,
    lesson.ClassNo,
    lesson.StartTime,
    lesson.Weeks,
    lesson.DayText,
    lesson.Venue,
  ].join('|');
}

/**
 * Check that two timetables are the same
 */
export function expectLessonsEqual(actual: RawLesson[], expected: RawLesson[]) {
  // Sort both expected and actual lessons because Jest expects
  // array to be in the same order
  expect(sortBy(actual, serializeLesson)).toEqual(sortBy(expected, serializeLesson));
}

function expectHistoryEqual(actual: SemesterData[], expected: SemesterData[]) {
  zip(
    sortBy(actual, (semester) => semester.Semester),
    sortBy(expected, (semester) => semester.Semester),
  ).forEach(([actualSemester, expectedSemester]) => {
    if (!actualSemester || !expectedSemester) throw new Error('Length of semesters not equal');

    expect(omit(actualSemester, 'Timetable')).toEqual(omit(expectedSemester, 'Timetable'));

    expectLessonsEqual(actualSemester.Timetable, expectedSemester.Timetable);
  });
}

export function expectModulesEqual(actual: Module, expected: Module) {
  // FulfillRequirements is excluded because the modules that require CS2100 are not part of this test
  const omittedKeys = ['SemesterData', 'FulfillRequirements'];
  expect(omit(actual, omittedKeys)).toEqual(omit(expected, omittedKeys));

  // Sort semesters and check history
  expectHistoryEqual(actual.SemesterData, expected.SemesterData);
}

/**
 * Mock an Axios response object
 */
export function mockResponse<T>(
  data: T,
  additionalFields: {
    headers?: { [name: string]: string };
    config?: Partial<AxiosRequestConfig>;
    request?: any;
    status?: number;
    statusText?: string;
  } = {},
): AxiosResponse<T> {
  const { headers, config, request, status, statusText } = additionalFields;

  return {
    data,
    // The server almost always returns 200, even if there is an application error
    status: status || 200,
    // @ts-ignore TS won't recognize this as valid because httpStatus is defined as a dictionary of static values
    statusText: statusText || httpStatus[status] || 'OK',
    headers: headers || {},
    config: config || {},
    request: request || {},
  };
}

/**
 * Helper to easily create sets less verbosely
 */
export function s<T>(...args: T[]): Set<T> {
  return new Set(args);
}
