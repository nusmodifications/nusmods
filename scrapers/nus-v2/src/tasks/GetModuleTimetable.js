// @flow
import { groupBy, partition, values } from 'lodash';
import NUSModerator from 'nusmoderator';

import type { ModuleCode, RawLesson, Semester, WeekText } from '../types/modules';
import type { Task } from '../types/tasks';

import BaseTask from './BaseTask';
import config from '../config';
import { getTermCode, retry } from '../utils/api';
import { validateLesson } from '../services/validation';
import type { TimetableLesson } from '../types/api';
import { activityLessonType, dayTextMap } from '../services/data';

/**
 * For deduplicating timetable lessons
 */
const getLessonKey = (lesson: TimetableLesson) =>
  [
    lesson.activity,
    lesson.modgrp,
    lesson.day,
    lesson.start_time,
    lesson.end_time,
    lesson.session,
    lesson.room,
    // '|' is used as a delimiter because it is unlikely to appear
    // organically in the data
  ].join('|');

/**
 * Try to infer week text from the provided list of events
 */
function getWeekText(lessons: TimetableLesson[]): WeekText {
  // All 13 weeks
  if (lessons.length === 13) return 'Every Week';

  // Get the week numbers the dates are in
  const weeks = lessons
    .map((lesson) => new Date(lesson.eventdate))
    .map((date) => NUSModerator.academicCalendar.getAcadWeekInfo(date))
    .map((weekInfo) => weekInfo.num)
    .sort((a, b) => a - b);

  // Calculate the number of weeks between lessons to check for
  // odd/even weeks
  const weekDelta = [];
  for (let i = 0; i < weeks.length - 1; i++) {
    weekDelta.push(weeks[i + 1] - weeks[i]);
  }

  if (weekDelta.every((delta) => delta === 2)) {
    // TODO: Check for tutorial / lab, those may only 5 classes
    if (weeks.length === 6) {
      return weeks[0] === 1 ? 'Odd Weeks' : 'Even Weeks';
    }
  }

  return weeks.join(',');
}

/**
 * Convert API provided timetable data to RawLesson format used by the frontend
 *
 * Output:
 *  - <semester>/<module code>/timetable.json
 */
export function mapTimetableLessons(lessons: TimetableLesson[]): RawLesson[] {
  // Group the same lessons together
  const groupedLessons = groupBy(lessons, getLessonKey);

  // For each lesson, map the keys from the NUS API to ours. Most have close
  // mappings, but week text needs to be inferred from the event's dates
  return values(groupedLessons).map((events: TimetableLesson[]) => {
    // eslint-disable-next-line camelcase
    const { room, start_time, end_time, day, modgrp, activity } = events[0];

    return {
      // mod group contains the activity at the start - we remove that because
      // it is redundant
      ClassNo: modgrp.replace(activity, ''),
      // Start and end time don't have the ':' delimiter
      StartTime: start_time.replace(':', ''),
      EndTime: end_time.replace(':', ''),
      // Week text is inferred from the event's dates
      WeekText: getWeekText(events),
      // Room can be null
      Venue: room || '',
      DayText: dayTextMap[day],
      LessonType: activityLessonType[activity],
    };
  });
}

/**
 * Download the timetable for a specific module
 */
export default class GetModuleTimetable extends BaseTask implements Task<void, RawLesson[]> {
  semester: Semester;
  academicYear: string;
  moduleCode: ModuleCode;

  get name() {
    return `Get timetable for ${this.moduleCode} for semester ${this.semester}`;
  }

  constructor(
    moduleCode: ModuleCode,
    semester: Semester,
    academicYear: string = config.academicYear,
  ) {
    super();

    this.semester = semester;
    this.academicYear = academicYear;
    this.moduleCode = moduleCode;

    this.logger = this.rootLogger.child({
      semester,
      moduleCode,
      task: GetModuleTimetable.name,
      year: academicYear,
    });
  }

  async run() {
    const term = getTermCode(this.semester, this.academicYear);

    const lessons = await retry(() => this.api.getModuleTimetable(term, this.moduleCode), 3);

    // Validate and remove invalid lessons
    const [validLessons, invalidLessons] = partition(lessons, validateLesson);
    if (invalidLessons.length > 0) {
      this.logger.info({ invalidLessons }, 'Removed %i invalid lessons', invalidLessons.length);
    }

    const timetable = mapTimetableLessons(validLessons);

    // Cache timetable to disk
    await this.output.timetable(this.semester, this.moduleCode, timetable);

    return timetable;
  }
}
