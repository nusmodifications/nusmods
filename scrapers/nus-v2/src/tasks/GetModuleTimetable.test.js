// @flow

import GetModuleTimetable from './GetModuleTimetable';
import { makeMockFile } from '../utils/test-utils';

describe(GetModuleTimetable, () => {
  test('should not crash if the data is invalid', async () => {
    const task = new GetModuleTimetable('CS1010', 1, '2018/2019');
    const lessons: any = [
      // Invalid lesson
      {
        term: '1810',
        room: null,
        numweeks: 13,
        start_time: '00:00',
        activity: 'L',
        csize: 40,
        module: 'CS1010',
        eventdate: null,
        session: '1',
        end_time: '00:00',
        modgrp: 'L1',
        deptfac: '00602ACAD1',
        day: null,
      },
    ];

    const mockFile = makeMockFile([]);
    task.api.getModuleTimetable = jest.fn().mockResolvedValue(lessons);
    task.fs.output.timetable = jest.fn().mockReturnValue(mockFile);

    const results = await task.run();

    expect(results).toEqual([]);
  });
  test('should filter out invalid lessons', async () => {
    const task = new GetModuleTimetable('CS1010', 1, '2018/2019');
    const lessons: any = [
      // Invalid lesson
      {
        term: '1810',
        room: null,
        numweeks: 13,
        start_time: '00:00',
        activity: 'L',
        csize: 40,
        module: 'CS1010',
        eventdate: null,
        session: '1',
        end_time: '00:00',
        modgrp: 'L1',
        deptfac: '00602ACAD1',
        day: null,
      },
      // Valid lesson
      {
        term: '1810',
        room: 'S16-0436',
        numweeks: 13,
        start_time: '14:00',
        activity: 'R',
        csize: 50,
        module: 'CS1010',
        eventdate: '2018-08-24',
        session: '1',
        end_time: '15:00',
        modgrp: 'R05',
        deptfac: '00301ACAD1',
        day: '5',
      },
    ];

    const mockFile = makeMockFile([]);
    task.api.getModuleTimetable = jest.fn().mockResolvedValue(lessons);
    task.fs.output.timetable = jest.fn().mockReturnValue(mockFile);

    const results = await task.run();

    expect(results).toMatchSnapshot();
    expect(mockFile.write).toHaveBeenCalled();
  });
});
