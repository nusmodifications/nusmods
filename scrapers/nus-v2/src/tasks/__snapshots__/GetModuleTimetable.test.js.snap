// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetModuleTimetable should filter out invalid lessons 1`] = `
Array [
  Object {
    "ClassNo": "05",
    "DayText": "Friday",
    "EndTime": "1500",
    "LessonType": "Recitation",
    "StartTime": "1400",
    "Venue": "S16-0436",
    "WeekText": "2",
  },
]
`;
